ionic-team/stencil
# feat(compiler): add `defineCustomElements` method & signature typedef

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features) - https://github.com/ionic-team/stencil-site/pull/921

## Pull request type

 

 

Please check the type of change your PR introduces:


## What is the current behavior?

`dist-custom-elements` currently does not have a `defineCustomElements` function to quickly register all components on the custom element registry.

GitHub Issue Number: [#3136](https://github.com/ionic-team/stencil/issues/3136#issuecomment-1130718870)


## What is the new behavior?


`bundle` is added as a `customElementsExportBehavior` option. This behavior will mirror that of `dist-custom-elements-bundle` in exporting a `defineCustomElements` function. This is intended to be used as a quick migration for users currently building with `dist-custom-elements-bundle`.

## Does this introduce a breaking change?




## Testing



Steps taken for testing:
1. Create a Stencil component starter
2. Build and pack/link this branch and use within the starter created above
3. In the starter, make sure `dist-custom-elements` is an output target and set `customElementsExportBehavior: 'bundle'`
4. Also set the appropriate targets in the stater's `package.json` according to the [Stencil documentation](https://stenciljs.com/docs/custom-elements#distributing-custom-elements) for the custom elements target.
5. Build and pack the starter app output
6. Add the packaged output as a dep to a starter framework project (I used Angular)
7. Import the `defineCustomElements()` function and call it on app bootstrap
8. Verify that any components in the component lib can be used in HTML

Unit tests have been added to the output target spec file as well as the type generation spec file.

## Other information
